tsl/src/nodes/decompress_chunk/decompress_chunk.c

```diff
@@ -359,6 +374,184 @@ cost_decompress_chunk(Path *path, Path *compressed_path)
+ * MergeAppend does, but this is not needed in our case and we could reduce the size of the heap and
+ * the amount of parallel open batches.
+ *
+ * The algorithm works as follows:
+ *
+ *   (1) A sort node is placed below the decompress scan node and on top of the scan
+ *       on the compressed chunk. This sort node uses the min/max values of the 'order by'
+ *       columns from the metadata of the batch to get them into an order which can be
+ *       used to merge them.
+ *
+ *       [Scan on compressed chunk] -> [Sort on min/max values] -> [Decompress and merge]
+ *
+ *       For example, the batches are sorted on the min value of the 'order by' metadata
+ *       column: [0, 3] [0, 5] [3, 7] [6, 10]
+ *
+ *   (2) The decompress chunk node initializes a binary heap, opens the first batch and
+ *       decompresses the first tuple from the batch. The tuple is put on the heap. In addition
+ *       the opened batch is marked as the most recent batch (MRB).
+ *
+ *   (3) As soon as a tuple is requested from the heap, the following steps are performed:
+ *       (3a) If the heap is empty, we are done.
+ *       (3b) The top tuple from the heap is taken. It is checked if this tuple is from the
+ *            MRB. If this is the case, the next batch is opened, the top tuple is decompressed,
+ *            placed on the batch, and this batch is marked as MRB. This is repeated until the
+ *            top tuple from the heap is not from the MRB. This is done to ensure we have opened
+ *            all relevant batches (and one ahead) which might contain the most recent tuple
+ *            are opened and placed on the heap.
+ *
+ *            In the example above, the first three batches are opened because the first two
+ *            batches might contain tuples with a value of 0.
+ *       (3c) The top element from the heap is removed, the next tuple from the batch is
+ *            decompressed (if present) and placed on the heap.
+ *       (3d) The former top tuple of the heap is returned.
+ *
+ * This function checks if the compression 'order by' and the query 'order by' are
+ * compatible and the optimization can be used.
+ */
+static MergeBatchResult
+is_able_to_use_segment_merge_append(PlannerInfo *root, CompressionInfo *info, Chunk *chunk)
+{
+	int pk_index;
+	PathKey *pk;
+	Var *var;
+	Expr *expr;
+	char *column_name;
+	List *pathkeys = root->query_pathkeys;
+	FormData_hypertable_compression *ci;
+	ListCell *lc = list_head(pathkeys);
+	MergeBatchResult merge_result = SCAN_FORWARD;
+
+	/* Ensure that we have path keys and the chunk is ordered */
+	if (pathkeys == NIL || ts_chunk_is_unordered(chunk) || ts_chunk_is_partial(chunk))
+		return MERGE_NOT_POSSIBLE;
+
+	/*
+	 * Loop over the of pathkeys of the query. These pathkeys
+	 * need to exactly match the configured compress_orderby pathkeys.
+	 */
+	for (pk_index = 1; lc != NULL; lc = lnext_compat(pathkeys, lc), pk_index++)
```

Comment:
PG Lists are arrays since a long time ago, so it's OK to use the more usual form with `list_length` and `list_nth`.

Reply:
Changed